internal/strategies/flexible_rollout.go

```diff
@@ -0,0 +1,79 @@
+package strategies
+
+import (
+	"math/rand"
+	"strconv"
+	"time"
+
+	"github.com/Unleash/unleash-client-go/v3/context"
+	"github.com/Unleash/unleash-client-go/v3/strategy"
+)
+
+type stickiness string
+
+const (
+	defaultStickiness   stickiness = "default"
+	userIDStickiness    stickiness = "userId"
+	sessionIDStickiness stickiness = "sessionId"
+	randomStickiness    stickiness = "random"
+)
+
+type flexibleRolloutStrategy struct {
+	random *rand.Rand
+}
+
+// NewFlexibleRolloutStrategy creates a new instance of the flexible rollout strategy.
+func NewFlexibleRolloutStrategy() *flexibleRolloutStrategy {
+	s := &flexibleRolloutStrategy{
+		random: rand.New(rand.NewSource(time.Now().Unix())),
+	}
+	return s
+}
+
+func (s flexibleRolloutStrategy) Name() string {
+	return "flexibleRollout"
+}
+
+func (s flexibleRolloutStrategy) randomNumber() string {
+	return strconv.Itoa(s.random.Intn(100) + 1)
+}
+
+func (s flexibleRolloutStrategy) resolveStickiness(st stickiness, ctx context.Context) string {
+	switch st {
+	case userIDStickiness:
+		return ctx.UserId
+	case sessionIDStickiness:
+		return ctx.SessionId
+	case randomStickiness:
+		return s.randomNumber()
+	default:
+		return coalesce(ctx.UserId, ctx.SessionId, s.randomNumber())
```

Comment:
Nice, I did not now that `coalesce` was a thing!

Reply:
It's not builtin unfortunately so I borrowed the concept from Postgres and implemented it in the "helpers" file. I think I might want to make a public API for these helpers since it makes writing custom strategies API. Especially things like `normalizedValue` are very helpful so this has already been requested by one user.

Reply:
Yes indeed. I'm planning the same thing for other clients.